pkg/syncer/admissionhandler/validatepvc.go

```diff
@@ -0,0 +1,198 @@
+package admissionhandler
+
+import (
+	"context"
+	"encoding/json"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v4/apis/volumesnapshot/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/v2/pkg/kubernetes"
+)
+
+const (
+	ExpandVolumeWithSnapshotErrorMessage = "Expanding volume with snapshots is not allowed"
+	DeleteVolumeWithSnapshotErrorMessage = "Deleting volume with snapshots is not allowed"
+)
+
+// validatePVC helps validate AdmissionReview requests for PersistentVolumeClaim.
+func validatePVC(ctx context.Context, ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	if containerOrchestratorUtility != nil && !containerOrchestratorUtility.IsFSSEnabled(ctx, common.BlockVolumeSnapshot) {
+		// If CSI block volume snapshot is disabled and webhook is running,
+		// skip validation for PersistentVolumeClaim.
+		return &admissionv1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	if ar.Request.Operation != admissionv1.Update && ar.Request.Operation != admissionv1.Delete {
+		// If AdmissionReview request operation is out of expectation,
+		// skip validation for PersistentVolumeClaim.
+		return &admissionv1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	log := logger.GetLogger(ctx)
+	req := ar.Request
+	var result *metav1.Status
+	allowed := true
+
+	switch req.Kind.Kind {
+	case "PersistentVolumeClaim":
+		oldPVC := corev1.PersistentVolumeClaim{}
+		log.Debugf("JSON req.OldObject.Raw: %v", string(req.OldObject.Raw))
+		if err := json.Unmarshal(req.OldObject.Raw, &oldPVC); err != nil {
+			log.Warnf("error deserializing old pvc: %v. skipping validation.", err)
+			return &admissionv1.AdmissionResponse{
+				// skip validation if there is pvc deserialization error
+				Allowed: true,
+			}
+		}
+		oldReq := oldPVC.Spec.Resources.Requests[corev1.ResourceStorage]
+
+		if !isRWOVolumeRequest(oldPVC.Spec.AccessModes) {
+			log.Info("the access mode of PVC is not ReadWriteOnce. skipping validation.")
+			return &admissionv1.AdmissionResponse{
+				// skip validation if the pvc is not RWO
+				Allowed: true,
+			}
+		}
+
```

Comment:
Do we need to check if this is a CSI migrated volume?

Reply:
The current design doesn't validate if the source volume is CSI migrated volume. We can add that validation in the CreateSnapshot case.

Here, we should not. Because here are validations for DeleteVolume and ExpandVolume. If we do any thing to a CSI migrated volume, it would be a change of behavior which is not expected.